x-pack/platform/plugins/shared/streams/server/lib/streams/client.ts

```diff
@@ -207,26 +215,126 @@ export class StreamsClient {
 
     const { dashboards, queries } = request;
 
-    const queryLinks = queries.map((query) => ({
-      [ASSET_ID]: query.id,
-      [ASSET_TYPE]: 'query' as const,
-      query,
-    }));
+    const { deleted, indexed, updated } = await this.dependencies.assetClient.syncAssetList(
+      stream.name,
+      [
+        ...dashboards.map((dashboard) => ({
+          [ASSET_ID]: dashboard,
+          [ASSET_TYPE]: 'dashboard' as const,
+        })),
+        ...queries.map((query) => ({
+          [ASSET_ID]: query.id,
+          [ASSET_TYPE]: 'query' as const,
+          query,
+        })),
+      ]
+    );
 
-    await this.dependencies.assetClient.syncAssetList(stream.name, [
-      ...dashboards.map((dashboard) => ({
-        [ASSET_ID]: dashboard,
-        [ASSET_TYPE]: 'dashboard' as const,
-      })),
-      ...queryLinks,
-    ]);
+    await this.manageQueries(name, {
+      deleted: deleted.filter(isQueryLink),
+      indexed: indexed.filter(isQueryLink),
+      updated: updated.filter(isQueryLink),
+    });
 
     return {
       acknowledged: true,
       result: result.changes.created.includes(name) ? 'created' : 'updated',
     };
   }
 
+  async manageQueries(
+    name: string,
+    {
+      deleted = [],
+      indexed = [],
+      updated = [],
+    }: { deleted?: QueryLink[]; indexed?: QueryLink[]; updated?: QueryLink[] }
+  ) {
+    if (this.dependencies.config.experimental?.significantEventsEnabled !== true) {
+      return;
+    }
+
+    await Promise.all([
+      this.installQueries(name, { indexed, updated }),
+      this.uninstallQueries(name, deleted),
+    ]);
+  }
+
+  private async installQueries(
+    name: string,
+    { indexed = [], updated = [] }: { indexed: QueryLink[]; updated: QueryLink[] }
+  ) {
+    const { rulesClient } = this.dependencies;
+
+    const indexedIds = new Set(indexed.map(getRuleIdFromQueryLink));
+    const updatedIds = new Set(updated.map(getRuleIdFromQueryLink));
+
+    const toUninstall = updated.filter((query) => !indexedIds.has(getRuleIdFromQueryLink(query)));
+    const toUpdate = indexed.filter((query) => updatedIds.has(getRuleIdFromQueryLink(query)));
+    const toCreate = indexed.filter((query) => !updatedIds.has(getRuleIdFromQueryLink(query)));
+
+    await Promise.all([
```

Comment:
I think we should harden this process in a follow up:
- pLimiter
- handle errors, log and continue

Reply:
Are there bulk APIs for this?

Reply:
no